### PR TITLE
aggregatedapiserver support openapi

### PR DESCRIPTION
**Pull request by carlory**
Signed-off-by: carlory <baofa.fan@daocloud.io>

**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

As #1122 merged, cluster crd has been moved to aggregation apiserver. 

```shell
(⎈ |karmada-apiserver:default)➜  karmada git:(openapi) kubectl explain cluster
error: couldn't find resource for "cluster.karmada.io/v1alpha1, Kind=Cluster"
```

expected:

```shell
(⎈ |karmada-apiserver:default)➜  karmada git:(openapi) kubectl explain cluster.spec
KIND:     Cluster
VERSION:  cluster.karmada.io/v1alpha1

RESOURCE: spec <Object>

DESCRIPTION:
     Spec represents the specification of the desired behavior of member
     cluster.

     ClusterSpec defines the desired state of a member cluster.

FIELDS:
   apiEndpoint	<string>
     The API endpoint of the member cluster. This can be a hostname,
     hostname:port, IP or IP:port.

   impersonatorSecretRef	<Object>
     ImpersonatorSecretRef represents the secret contains the token of
     impersonator. The secret should hold credentials as follows: -
     secret.data.token

   insecureSkipTLSVerification	<boolean>
     InsecureSkipTLSVerification indicates that the karmada control plane should
     not confirm the validity of the serving certificate of the cluster it is
     connecting to. This will make the HTTPS connection between the karmada
     control plane and the member cluster insecure. Defaults to false.

   provider	<string>
     Provider represents the cloud provider name of the member cluster.

   proxyURL	<string>
     ProxyURL is the proxy URL for the cluster. If not empty, the karmada
     control plane will use this proxy to talk to the cluster. More details
     please refer to: https://github.com/kubernetes/client-go/issues/351

   region	<string>
     Region represents the region of the member cluster locate in.

   secretRef	<Object>
     SecretRef represents the secret contains mandatory credentials to access
     the member cluster. The secret should hold credentials as follows: -
     secret.data.token - secret.data.caBundle

   syncMode	<string> -required-
     SyncMode describes how a cluster sync resources from karmada control plane.

   taints	<[]Object>
     Taints attached to the member cluster. Taints on the cluster have the
     "effect" on any resource that does not tolerate the Taint.

   zone	<string>
     Zone represents the zone of the member cluster locate in.
```

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: examples/customresourceinterpreter/apis/workload/v1alpha1/zz_generated.deepcopy.go [hunk 1]
```diff
@@ -1,3 +1,4 @@
+//go:build !ignore_autogenerated
```
Comment (XiShanYongYe-Chang) [inline examples/customresourceinterpreter/apis/workload/v1alpha1/zz_generated.deepcopy.go:1]: @RainbowMango This line has been added again. :>

File: hack/update-codegen.sh [hunk 2]
```diff
@@ -114,3 +114,13 @@ informer-gen \
...
   --listers-package=github.com/karmada-io/karmada/pkg/generated/listers \
   --output-package=github.com/karmada-io/karmada/pkg/generated/informers
+
+echo "Generating with openapi-gen"
+GO111MODULE=on go install k8s.io/code-generator/cmd/openapi-gen
+openapi-gen \
+  --go-header-file hack/boilerplate/boilerplate.go.txt \
+  --input-dirs "github.com/karmada-io/karmada/pkg/apis/cluster/v1alpha1" \
+  --input-dirs "k8s.io/api/core/v1,k8s.io/apimachinery/pkg/api/resource" \
+  --input-dirs "k8s.io/apimachinery/pkg/apis/meta/v1,k8s.io/apimachinery/pkg/runtime,k8s.io/apimachinery/pkg/version" \
```
Comment (XiShanYongYe-Chang) [inline hack/update-codegen.sh:124]: Is it directly written as fixed here? I've seen a recursive way of generating.

File: examples/customresourceinterpreter/apis/workload/v1alpha1/zz_generated.deepcopy.go (hunk 1 above)
Comment (carlory) [reply examples/customresourceinterpreter/apis/workload/v1alpha1/zz_generated.deepcopy.go:1]: @XiShanYongYe-Chang @RainbowMango 

`kubernetes/kubernetes` also has the same situation.

https://github.com/kubernetes/api/blob/master/apps/v1/zz_generated.deepcopy.go#L2

File: hack/update-codegen.sh (hunk 2 above)
Comment (carlory) [reply hack/update-codegen.sh:124]: @XiShanYongYe-Chang 

We must specify the dependencies as fixed here. without it, the karmada-aggregated-apiserver will fail to start.

```shell
➜  ~ kubectl -n karmada-system logs karmada-aggregated-apiserver-b8876cfdb-s4pgt
I0104 05:41:41.811682       1 plugins.go:158] Loaded 2 mutating admission controller(s) successfully in the following order: NamespaceLifecycle,MutatingAdmissionWebhook.
I0104 05:41:41.811800       1 plugins.go:161] Loaded 1 validating admission controller(s) successfully in the following order: ValidatingAdmissionWebhook.
Error: unable to get openapi models: cannot find model definition for k8s.io/apimachinery/pkg/api/resource.Quantity. If you added a new type, you may need to add +k8s:openapi-gen=true to the package or type and run code-gen again
Usage:
  karmada-aggregated-apiserver [flags]
  karmada-aggregated-apiserver [command]

...
```

Comment (XiShanYongYe-Chang) [reply hack/update-codegen.sh:124]: code generator provides a script `generate-internal-groups.sh` to generate internal groups, can we use it?

https://github.com/kubernetes/code-generator/blob/05a01c0d31e8d2d8ecc63342181c8fe302d1c04c/generate-internal-groups.sh#L21-L22

File: examples/customresourceinterpreter/apis/workload/v1alpha1/zz_generated.deepcopy.go (hunk 1 above)
Comment (XiShanYongYe-Chang) [reply examples/customresourceinterpreter/apis/workload/v1alpha1/zz_generated.deepcopy.go:1]: I'm not saying it's wrong. I've added this before.

File: hack/update-codegen.sh (hunk 2 above)
Comment (carlory) [reply hack/update-codegen.sh:124]: @XiShanYongYe-Chang we can use it, but it still requires that we specify the missing dependencies. for example, `k8s.io/apimachinery/pkg/api/resource`

File: examples/customresourceinterpreter/apis/workload/v1alpha1/zz_generated.deepcopy.go (hunk 1 above)
Comment (RainbowMango) [reply examples/customresourceinterpreter/apis/workload/v1alpha1/zz_generated.deepcopy.go:1]: It's time to figure out why and how this change is introduced.

**Comment by RainbowMango** [comment]
/hold
for waiting #1220

File: examples/customresourceinterpreter/apis/workload/v1alpha1/zz_generated.deepcopy.go (hunk 1 above)
Comment (XiShanYongYe-Chang) [reply examples/customresourceinterpreter/apis/workload/v1alpha1/zz_generated.deepcopy.go:1]: Hi, @carlory, which go version do you use?

**Comment by RainbowMango** [comment]
/hold cancel
As #1220 has been solved.

@carlory Can you rebase your code and push again?

**Comment by carlory** [comment]
@RainbowMango can you give me some advice about it?

```shell
➜  karmada git:(openapi) git push -f
Everything up-to-date
➜  karmada git:(openapi) hack/verify-codegen.sh
Generating with deepcopy-gen
Generating with register-gen
Generating with conversion-gen
Generating with client-gen
Generating with lister-gen
Generating with informer-gen
Generating with openapi-gen
API rule violation: list_type_missing,github.com/karmada-io/karmada/pkg/apis/cluster/v1alpha1,APIEnablement,Resources
...
diffing hack/../pkg against freshly generated codegen
hack/../pkg up to date.
➜  karmada git:(openapi) git status
On branch openapi
Your branch is up to date with 'origin/openapi'.

nothing to commit, working tree clean
```

**Comment by carlory** [comment]
@RainbowMango @iawia002 cloud you give me some guide about it?

```shell
➜  karmada git:(openapi) ✗ hack/verify-codegen.sh
Generating with deepcopy-gen
Generating with register-gen
Generating with conversion-gen
Generating with client-gen
Generating with lister-gen
Generating with informer-gen
Generating with openapi-gen
API rule violation: list_type_missing,github.com/karmada-io/karmada/pkg/apis/cluster/v1alpha1,APIEnablement,Resources
...
diffing hack/../pkg against freshly generated codegen
hack/../pkg up to date.
```

**Comment by iawia002** [comment]
```diff
+//go:build !ignore_autogenerated
 // +build !ignore_autogenerated
```

I think this is a problem with the go version like @XiShanYongYe-Chang mentioned at https://github.com/karmada-io/karmada/pull/1201#discussion_r780056032, we should use the same go version as in CI to generate the code.

**Comment by RainbowMango** [comment]
Yes. We should use Go v1.16, as from v1.17, the `// +build` will be replaced by `go:build`.

**Comment by RainbowMango** [comment]
You can remove all the generated files locally, then generates the files with Go 1.16.x.

**Comment by XiShanYongYe-Chang** [comment]
More detail for `go:build`: https://go.dev/doc/go1.17#gofmt

File: pkg/apis/work/v1alpha2/zz_generated.deepcopy.go [hunk 3]
```diff
@@ -31,6 +31,27 @@ func (in *AggregatedStatusItem) DeepCopy() *AggregatedStatusItem {
 	return out
 }
 
+// DeepCopyInto is an autogenerated deepcopy function, copying the receiver, writing into out. in must be non-nil.
```
Comment (iawia002) [inline pkg/apis/work/v1alpha2/zz_generated.deepcopy.go:34]: This is already updated in fc20a6c040239a30f9f5467e87a85bf91bda0486

Comment (carlory) [reply pkg/apis/work/v1alpha2/zz_generated.deepcopy.go]: @iawia002 thanks, it is fixed.

**Comment by RainbowMango** [comment]
Found a failed case:
```
[MCS] Multi-Cluster Service testing EndpointSlices change testing 
  Update Deployment's replicas
  /home/runner/work/karmada/karmada/test/e2e/mcs_test.go:403
STEP: Creating ClusterPropagationPolicy(serviceexports-pb7-policy)
STEP: Creating ClusterPropagationPolicy(serviceimports-czg-policy)
STEP: Check if crd(multicluster.x-k8s.io/v1alpha1/ServiceExport) present on member clusters
I0110 03:04:02.880409   47350 customresourcedefine.go:58] Waiting for crd present on cluster(member1)
I0110 03:04:02.910309   47350 customresourcedefine.go:58] Waiting for crd present on cluster(member2)
I0110 03:04:03.001134   47350 customresourcedefine.go:58] Waiting for crd present on cluster(member3)
STEP: Check if crd(multicluster.x-k8s.io/v1alpha1/ServiceImport) present on member clusters
I0110 03:04:03.046667   47350 customresourcedefine.go:58] Waiting for crd present on cluster(member1)
I0110 03:04:03.067802   47350 customresourcedefine.go:58] Waiting for crd present on cluster(member2)
I0110 03:04:03.092907   47350 customresourcedefine.go:58] Waiting for crd present on cluster(member3)
STEP: Creating Deployment(karmadatest-6ch/hello-nnl)
I0110 03:04:03.120935   47350 mcs_test.go:370] Create Deployment(karmadatest-6ch/hello-nnl) in member1 cluster
STEP: Creating Service(karmadatest-6ch/hello-nnl)
I0110 03:04:03.144055   47350 mcs_test.go:373] Create Service(karmadatest-6ch/hello-nnl) in member1 cluster
STEP: Wait Service(karmadatest-6ch/hello-nnl)'s EndpointSlice exist in member1 cluster
STEP: Create ServiceExport(karmadatest-6ch/hello-nnl)
STEP: Creating PropagationPolicy(karmadatest-6ch/export-hello-nnl-policy)
STEP: Wait EndpointSlices collected to namespace(karmadatest-6ch) in controller-plane
STEP: Removing Deployment(karmadatest-6ch/hello-nnl)
I0110 03:09:08.241555   47350 mcs_test.go:396] Delete Deployment(karmadatest-6ch/hello-nnl) in member1 cluster
I0110 03:09:08.258929   47350 mcs_test.go:399] Delete Service(karmadatest-6ch/hello-nnl) in member1 cluster
STEP: Removing Service(karmadatest-6ch/hello-nnl)
STEP: Removing ClusterPropagationPolicy(serviceimports-czg-policy)
STEP: Removing ClusterPropagationPolicy(serviceexports-pb7-policy)

• Failure [305.658 seconds]
[MCS] Multi-Cluster Service testing
/home/runner/work/karmada/karmada/test/e2e/mcs_test.go:178
  EndpointSlices change testing
  /home/runner/work/karmada/karmada/test/e2e/mcs_test.go:363
    Update Deployment's replicas [It]
    /home/runner/work/karmada/karmada/test/e2e/mcs_test.go:403

    Timed out after 300.003s.
    Expected
        <int>: 0
    to equal
        <int>: 1

    /home/runner/work/karmada/karmada/test/e2e/mcs_test.go:428
------------------------------
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS

Summarizing 1 Failure:

[Fail] [MCS] Multi-Cluster Service testing EndpointSlices change testing [It] Update Deployment's replicas 
/home/runner/work/karmada/karmada/test/e2e/mcs_test.go:428

Ran 3 of 39 Specs in 349.401 seconds
FAIL! -- 2 Passed | 1 Failed | 0 Pending | 36 Skipped
--- FAIL: TestE2E (349.55s)
FAIL
```

File: hack/update-codegen.sh (hunk 2 above)
Comment (RainbowMango) [reply hack/update-codegen.sh:124]: Where does `openapi-gen` gets the `--input-dirs`? From vendor or `GOPATH/src/`?

I'm thinking if it still works if the dependency packages do not exit in `GOPATH/src/`, that might happen in our CI system.

Comment (XiShanYongYe-Chang) [reply hack/update-codegen.sh:124]: I test it on my local, it gets from the vendor.

**Comment by XiShanYongYe-Chang** [comment]
> https://github.com/karmada-io/karmada/runs/4756599935?check_suite_focus=true

The failure cause is the same as #1195, as a result, the ServiceExport object can not be propagated to the member1 cluster.

**Pull request merged**